File: control/drone_course_ws/src/exercise_3/src/exercise_3.cpp
// Copyright 2026 Universidad Politécnica de Madrid
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <algorithm>
#include <cmath>
#include <limits>
#include <exercise_3/exercise_3.hpp>

namespace drone_course
{

  DroneCourseExercise3::DroneCourseExercise3(
      const std::string &node_name,
      const rclcpp::NodeOptions &options)
      : rclcpp::Node(node_name, options)
  {
    // QoS profiles
    rclcpp::QoS reliable_qos = rclcpp::QoS(10).reliable();
    rclcpp::QoS best_effort_qos = rclcpp::QoS(10).best_effort();

    callback_group_ =
        this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);

    // Suscribers

    pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
        "/drone0/self_localization/pose", best_effort_qos,
        std::bind(&DroneCourseExercise3::state_subscription_callback, this, std::placeholders::_1));

    // Publishers

    vel_pub_ = this->create_publisher<geometry_msgs::msg::TwistStamped>(
        "/drone0/motion_reference/twist", reliable_qos);

    // Services clients

    control_mode_service_client_ = this->create_client<as2_msgs::srv::SetControlMode>(
        "/drone0/controller/set_control_mode", rmw_qos_profile_services_default, callback_group_);

    path_service_client_ = this->create_client<drone_course_msgs::srv::RequestPath>(
        "/request_path", rmw_qos_profile_services_default, callback_group_);

    // Timers
    double timer_freq = 100.0;
    timer_ = this->create_wall_timer(
        std::chrono::duration<double>(1.0f / timer_freq),
        std::bind(&DroneCourseExercise3::timer_callback, this));

    dt_ = 1.0f / timer_freq;

    RCLCPP_INFO(this->get_logger(), "DroneCourseExercise3 initialized\n");
  }

  DroneCourseExercise3::~DroneCourseExercise3() {}

  std::vector<double> DroneCourseExercise3::solve_linear_system(
      std::vector<std::vector<double>> A, std::vector<double> b)
  {
    int N = static_cast<int>(b.size());
    for (int col = 0; col < N; col++)
    {
      int pivot = col;
      for (int row = col + 1; row < N; row++)
        if (std::abs(A[row][col]) > std::abs(A[pivot][col]))
          pivot = row;
      std::swap(A[col], A[pivot]);
      std::swap(b[col], b[pivot]);
      for (int row = col + 1; row < N; row++)
      {
        double f = A[row][col] / A[col][col];
        for (int k = col; k < N; k++)
          A[row][k] -= f * A[col][k];
        b[row] -= f * b[col];
      }
    }
    std::vector<double> x(N);
    for (int i = N - 1; i >= 0; i--)
    {
      x[i] = b[i];
      for (int j = i + 1; j < N; j++)
        x[i] -= A[i][j] * x[j];
      x[i] /= A[i][i];
    }
    return x;
  }

  void DroneCourseExercise3::build_spline()
  {
    int N = static_cast<int>(waypoints_.size());
    spline_M_.assign(N, {0.0, 0.0, 0.0});

    // C2 continuity
    std::vector<std::vector<double>> A(N, std::vector<double>(N, 0.0));
    for (int i = 0; i < N; i++)
    {
      A[i][i] = 4.0;
      A[i][(i + 1) % N] = 1.0;
      A[i][(i - 1 + N) % N] = 1.0;
    }

    for (int k = 0; k < 3; k++)
    {
      std::vector<double> d(N);
      for (int i = 0; i < N; i++)
      {
        d[i] = 6.0 * (waypoints_[(i + 1) % N][k] - 2.0 * waypoints_[i][k] + waypoints_[(i - 1 + N) % N][k]);
      }
      auto M = solve_linear_system(A, d);
      for (int i = 0; i < N; i++)
        spline_M_[i][k] = M[i];
    }
  }

  std::array<double, 3> DroneCourseExercise3::sample_spline(double t) const
  {
    int N = static_cast<int>(waypoints_.size());
    double t_mod = std::fmod(t, static_cast<double>(N));
    if (t_mod < 0.0)
      t_mod += N;

    int i = static_cast<int>(t_mod);
    double s = t_mod - i;
    int j = (i + 1) % N;

    std::array<double, 3> result;
    for (int k = 0; k < 3; k++)
    {
      double p_i = waypoints_[i][k];
      double p_j = waypoints_[j][k];
      double m_i = spline_M_[i][k];
      double m_j = spline_M_[j][k];
      result[k] = (1.0 - s) * p_i + s * p_j + (m_i * (1.0 - s) * ((1.0 - s) * (1.0 - s) - 1.0) + m_j * s * (s * s - 1.0)) / 6.0;
    }
    return result;
  }

  std::array<double, 3> DroneCourseExercise3::sample_spline_deriv(double t) const
  {
    int N = static_cast<int>(waypoints_.size());
    double t_mod = std::fmod(t, static_cast<double>(N));
    if (t_mod < 0.0)
      t_mod += N;

    int i = static_cast<int>(t_mod);
    double s = t_mod - i;
    int j = (i + 1) % N;

    std::array<double, 3> result;
    for (int k = 0; k < 3; k++)
    {
      double p_i = waypoints_[i][k];
      double p_j = waypoints_[j][k];
      double m_i = spline_M_[i][k];
      double m_j = spline_M_[j][k];
      result[k] = (p_j - p_i) + (m_i * (1.0 - 3.0 * (1.0 - s) * (1.0 - s)) + m_j * (3.0 * s * s - 1.0)) / 6.0;
    }
    return result;
  }

  double DroneCourseExercise3::find_closest_t(double x, double y, double z) const
  {
    int N = static_cast<int>(waypoints_.size());

    const int samples_per_seg = 50;
    const double step = 1.0 / samples_per_seg;

    double best_t = traj_t_;
    double best_dist_sq = std::numeric_limits<double>::max();

    for (int s = 0; s <= N * samples_per_seg; s++)
    {
      double t = traj_t_ + s * step;
      auto pt = sample_spline(t);
      double dx = pt[0] - x, dy = pt[1] - y, dz = pt[2] - z;
      double dist_sq = dx * dx + dy * dy + dz * dz;
      if (dist_sq < best_dist_sq)
      {
        best_dist_sq = dist_sq;
        best_t = t;
      }
    }
    return best_t;
  }

  double DroneCourseExercise3::find_lookahead_t(double t_proj, double lookahead) const
  {
    int N = static_cast<int>(waypoints_.size());
    const double dt_arc = 0.02;
    const double max_search = static_cast<double>(N);

    double t = t_proj;
    double arc = 0.0;

    while (arc < lookahead && (t - t_proj) < max_search)
    {
      auto p1 = sample_spline(t);
      auto p2 = sample_spline(t + dt_arc);
      double dx = p2[0] - p1[0], dy = p2[1] - p1[1], dz = p2[2] - p1[2];
      double seg_len = std::sqrt(dx * dx + dy * dy + dz * dz);

      if (arc + seg_len >= lookahead)
      {
        t += dt_arc * (lookahead - arc) / seg_len;
        break;
      }
      arc += seg_len;
      t += dt_arc;
    }
    return t;
  }

  void DroneCourseExercise3::timer_callback()
  {
    if (!path_received_)
    {
      path_service_request_ = std::make_shared<drone_course_msgs::srv::RequestPath::Request>();
      rclcpp::Client<drone_course_msgs::srv::RequestPath>::SharedFuture future =
          path_service_client_->async_send_request(path_service_request_).future.share();
      future.wait();
      std::vector<drone_course_msgs::msg::Point> path = future.get()->path;
      if (path.size() > 0)
      {
        RCLCPP_INFO(this->get_logger(), "Path received successfully (%zu waypoints)", path.size());
        waypoints_.clear();
        for (const auto &pt : path)
        {
          double x = static_cast<double>(pt.x) + (pt.x < 7.0f ? 1.0 : -1.0);
          waypoints_.push_back({x,
                                static_cast<double>(pt.y),
                                static_cast<double>(pt.z)});
        }
        build_spline();
        traj_ready_ = true;
        path_received_ = true;
      }
      else
      {
        RCLCPP_ERROR(this->get_logger(), "Failed to get path");
      }
    }

    // Check if control mode is set, if not, call service
    if (!control_mode_set_)
    {
      RCLCPP_INFO(this->get_logger(), "Calling control mode service...");
      auto control_mode_request = std::make_shared<as2_msgs::srv::SetControlMode::Request>();
      control_mode_request->control_mode.control_mode = as2_msgs::msg::ControlMode::SPEED;
      control_mode_request->control_mode.yaw_mode = as2_msgs::msg::ControlMode::YAW_SPEED;
      control_mode_request->control_mode.reference_frame =
          as2_msgs::msg::ControlMode::LOCAL_ENU_FRAME;

      rclcpp::Client<as2_msgs::srv::SetControlMode>::SharedFuture future =
          control_mode_service_client_->async_send_request(control_mode_request).future.share();
      future.wait();
      as2_msgs::srv::SetControlMode::Response::SharedPtr response = future.get();
      if (response->success)
      {
        RCLCPP_INFO(this->get_logger(), "Control mode set successfully");
        control_mode_set_ = true;
      }
      else
      {
        RCLCPP_ERROR(this->get_logger(), "Failed to set control mode");
      }
    }

    if (!traj_ready_ || !control_mode_set_)
      return;

    // Read current drone state
    double current_x = state_pose_.pose.position.x;
    double current_y = state_pose_.pose.position.y;
    double current_z = state_pose_.pose.position.z;

    traj_t_ = find_closest_t(current_x, current_y, current_z);
    double t_look = find_lookahead_t(traj_t_, lookahead_dist_);
    auto lookahead_pt = sample_spline(t_look);

    double dx = lookahead_pt[0] - current_x;
    double dy = lookahead_pt[1] - current_y;
    double dz = lookahead_pt[2] - current_z;

    double Kpp = desired_speed_ / lookahead_dist_;
    double velocity_x = Kpp * dx;
    double velocity_y = Kpp * dy;
    double velocity_z = Kpp * dz;

    double speed = std::sqrt(velocity_x * velocity_x + velocity_y * velocity_y + velocity_z * velocity_z);
    if (speed > desired_speed_)
    {
      double scale = desired_speed_ / speed;
      velocity_x *= scale;
      velocity_y *= scale;
      velocity_z *= scale;
    }

    // Publish velocity command
    geometry_msgs::msg::TwistStamped velocity_msg;
    velocity_msg.header.stamp = this->get_clock()->now();
    velocity_msg.header.frame_id = "earth";
    velocity_msg.twist.linear.x = velocity_x;
    velocity_msg.twist.linear.y = velocity_y;
    velocity_msg.twist.linear.z = velocity_z;
    velocity_msg.twist.angular.x = 0.0;
    velocity_msg.twist.angular.y = 0.0;
    velocity_msg.twist.angular.z = 0.0;

    if (vel_pub_)
    {
      vel_pub_->publish(velocity_msg);
    }
  }

  void DroneCourseExercise3::state_subscription_callback(
      const geometry_msgs::msg::PoseStamped::SharedPtr msg)
  {
    state_pose_ = *msg;
  }

} // namespace drone_course

File: vision/python_interface/exercise_3.py
# Copyright 2026 Universidad Politécnica de Madrid
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#
#    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
#      contributors may be used to endorse or promote products derived from
#      this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.


import argparse
import os

import cv2
import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import numpy as np
from scipy.spatial.transform import Rotation

from exercise_2 import detect_corners

Coord3D = tuple[float, float, float]
Quaternion = tuple[float, float, float, float]
Localization = tuple[Coord3D, Quaternion]

# Camera intrinsics (from dataset_info.yaml)
IMAGE_WIDTH = 1640
IMAGE_HEIGHT = 1232

K = np.array(
    [[365.77364921569824, 0.0, 820.0], [0.0, 365.773645401001, 616.0], [0.0, 0.0, 1.0]],
    dtype=np.float64,
)

# Distortion coefficients [k1, k2, p1, p2, k3] - plumb_bob model
DIST_COEFFS = np.zeros((5, 1), dtype=np.float64)  # all zero = no distortion


def load_dataset(dataset_path: str) -> list[tuple[cv2.Mat, list[Localization]]]:
    dataset = []
    images_path = os.path.join(dataset_path, 'images')
    if not os.path.exists(images_path):
        print(f'Images directory not found: {images_path}')
        return dataset

    labels_path = os.path.join(dataset_path, 'labels')
    if not os.path.exists(labels_path):
        print(f'Labels directory not found: {labels_path}')
        return dataset

    for filename in os.listdir(images_path):
        if filename.endswith('.jpg') or filename.endswith('.png'):
            image_path = os.path.join(images_path, filename)
            image: np.ndarray | None = cv2.imread(image_path)
            if image is None:
                print(f'Failed to load image: {image_path}')
                continue
            h, w = image.shape[:2]

            label_path = os.path.join(
                labels_path, filename.replace('.jpg', '.txt').replace('.png', '.txt')
            )
            labels: list[Localization] = []
            with open(label_path, 'r') as f:
                for line in f:
                    parts = line.strip().split()
                    parts = list(map(float, parts))
                    coords: Coord3D = (parts[0], parts[1], parts[2])
                    orient: Quaternion = (parts[3], parts[4], parts[5], parts[6])
                    labels.append((coords, orient))

            dataset.append((image, labels))

    return dataset


def translation_error(t1: Coord3D, t2: Coord3D) -> float:
    v1 = np.array(list(t1))
    v2 = np.array(list(t2))
    return float(np.linalg.norm(v1 - v2))


def rotation_error(o1: Quaternion, o2: Quaternion):
    q1 = np.array(o1)
    q2 = np.array(o2)

    q1 = q1 / np.linalg.norm(q1)
    q2 = q2 / np.linalg.norm(q2)

    dot = abs(np.dot(q1, q2))

    dot = np.clip(dot, 0.0, 1.0)

    angle_rad = 2.0 * np.arccos(dot)
    return np.degrees(angle_rad)


def localize_gate(image: np.ndarray) -> list[Localization]:
    # TODO (Exercise 3): Use the code from previous tasks to calculate the camera position relative
    # to the gate.
    h, w = image.shape[:2]

    corners = detect_corners(image)
    if len(corners) != 8:
        return [((0.0, 0.0, 0.0), (1.0, 0.0, 0.0, 0.0))]

    pts_2d = np.array([[cx * w, cy * h] for (cx, cy) in corners], dtype=np.float64)

    outer_gate_size = 2.7
    inner_gate_size = 1.5

    inner_top_y = -(outer_gate_size + inner_gate_size) / 2.0
    inner_bottom_y = -(outer_gate_size - inner_gate_size) / 2.0

    pts_3d = np.array([
        [-outer_gate_size / 2.0, -outer_gate_size, 0.0],  # TL outer 
        [ outer_gate_size / 2.0, -outer_gate_size, 0.0],  # TR outer 
        [-inner_gate_size / 2.0,  inner_top_y,     0.0],  # TL inner
        [ inner_gate_size / 2.0,  inner_top_y,     0.0],  # TR inner
        [-inner_gate_size / 2.0,  inner_bottom_y,  0.0],  # BL inner
        [ inner_gate_size / 2.0,  inner_bottom_y,  0.0],  # BR inner
        [-outer_gate_size / 2.0,  0.0,             0.0],  # BL outer 
        [ outer_gate_size / 2.0,  0.0,             0.0],  # BR outer 
    ], dtype=np.float64)

    success, rvec, tvec = cv2.solvePnP(pts_3d, pts_2d, K, DIST_COEFFS, flags=cv2.SOLVEPNP_SQPNP)
    if not success:
        return [((0.0, 0.0, 0.0), (1.0, 0.0, 0.0, 0.0))]

    R_obj_to_cam, _ = cv2.Rodrigues(rvec)
    t_obj_to_cam = tvec

    R_cam_to_obj_cv = R_obj_to_cam.T
    t_cam_to_obj_cv = -R_cam_to_obj_cv @ t_obj_to_cam

    R_cv_to_world = np.array([
        [ 0.0,  0.0,  1.0],
        [-1.0,  0.0,  0.0],
        [ 0.0, -1.0,  0.0]
    ])
    
    R_cam_to_cv_cam = np.array([
        [ 0.0, -1.0,  0.0],
        [ 0.0,  0.0, -1.0],
        [ 1.0,  0.0,  0.0]
    ])
    
    t_final_world = R_cv_to_world @ t_cam_to_obj_cv
    tx, ty, tz = t_final_world.flatten()

    R_final_camera_in_world = R_cv_to_world @ R_cam_to_obj_cv @ R_cam_to_cv_cam

    qx, qy, qz, qw = Rotation.from_matrix(R_final_camera_in_world).as_quat()
    
    return [((tx, ty, tz), (qw, qx, qy, qz))]


def parse_arguments():
    parser = argparse.ArgumentParser(
        description='Load a dataset of images and segmentation labels.'
    )
    parser.add_argument(
        '--dataset_path',
        type=str,
        help='Path to the dataset directory',
        default='../datasets/positions',
    )
    return parser.parse_args()


def visualize_3d(gt_pose, calc_pose):
    fig = plt.figure(figsize=(10, 8))
    ax = fig.add_subplot(111, projection='3d')

    s_out = 2.7
    gate_x = [0, 0, 0, 0, 0]
    gate_y = [s_out/2, -s_out/2, -s_out/2, s_out/2, s_out/2]
    gate_z = [2.7, 2.7, 0.0, 0.0, 2.7]
    ax.plot(gate_x, gate_y, gate_z, 'k-', linewidth=3, label='Gate')

    def draw_pose(pose, name, line_style):
        t, q = pose
        tx, ty, tz = t
        
        qw, qx, qy, qz = q
        R = Rotation.from_quat([qx, qy, qz, qw]).as_matrix()
        
        ax.quiver(tx, ty, tz, R[0,0], R[1,0], R[2,0], color='r', length=0.8, linestyle=line_style)
        ax.quiver(tx, ty, tz, R[0,1], R[1,1], R[2,1], color='g', length=0.8, linestyle=line_style)
        ax.quiver(tx, ty, tz, R[0,2], R[1,2], R[2,2], color='b', length=0.8, linestyle=line_style)
        
        ax.scatter(tx, ty, tz, color='k', s=40)
        ax.text(tx, ty, tz + 0.2, name, color='k', fontsize=10)

    draw_pose(gt_pose, "Ground Truth", '-')
    draw_pose(calc_pose, "Estimated", '--')

    ax.set_xlabel('Global X (Forward/Backward)')
    ax.set_ylabel('Global Y (Left/Right)')
    ax.set_zlabel('Global Z (Up/Down)')
    
    ax.set_box_aspect([1, 1, 1])
    
    ax.set_xlim([-4, 1])
    ax.set_ylim([-3, 3])  
    ax.set_zlim([0, 4])  
    
    plt.title("3D Global Coordinate Debugger")
    plt.legend()
    plt.show()

if __name__ == '__main__':
    args = parse_arguments()
    dataset = load_dataset(args.dataset_path)
    print(f'Loaded {len(dataset)} images and their corresponding labels.')

    result: list[tuple[np.ndarray, list[Localization], list[Localization]]] = []
    for image, labels in dataset:
        calculated_position: list[Localization] = localize_gate(image)
        result.append((image, labels, calculated_position))

    translation_errors: list[float] = []
    orientation_errors: list[float] = []

    for r in result:
        image: np.ndarray = r[0]
        gt_pose: Localization = r[1][0]
        calculated_pose: Localization = r[2][0]

        t_error = translation_error(gt_pose[0], calculated_pose[0])
        r_error = rotation_error(gt_pose[1], calculated_pose[1])

        translation_errors.append(t_error)
        orientation_errors.append(r_error)
        #visualize_3d(gt_pose, calculated_pose)
        #cv2.waitKey(0)

    print('Mean translation error = ', np.mean(translation_errors))
    print('Mean orientation error = ', np.mean(orientation_errors))
